src/cpp/thread_manager/thread_manager.cc

```diff
@@ -48,8 +45,14 @@ void ThreadManager::WorkerThread::Run() {
 }
 
 ThreadManager::WorkerThread::~WorkerThread() {
-  // Don't join until the thread is fully constructed.
-  std::lock_guard<std::mutex> lock(wt_mu_);
+  // the thread will always be fully constructed by now because the destructor
+  // is only ever called if:
+  //    1. valid_ was false (in which case an asynchronous thread never ran)
```

Comment:
I am not sure if this statement is true.. 
     `valid_ = thd_mgr->thread_creator_(&thd_, ...)`
Isn't it possible that the thread was created by `thread_creator_` call but has not updated `thd` yet (and/or valid_ is not updated yet) but the thread finished execution??

Reply:
So the only case where a WorkerThread gets deleted directly (without going through CleanupCompletedThreads) is if the thread creation failed, in which case no thread ever ran and `valid_` is false. I can change the comment to express it in that fashion, if you think that's appropriate.